### ISSUE TITLE
Add MPI support

### ISSUE THREAD

**Issue by mikebentley15** [PRUNERS/FLiT#127]
Many potential users write their code with the MPI libraries.  I believe all that is necessary to complete support for MPI is to add a wrapper variable in the Makefile, say called `RUN_WRAPPER`, that can be overridden in `custom.mk`.  I think all that would be needed is to set

```
RUN_WRAPPER = mpirun -n 4
```

within `custom.mk`.  This obviously needs to be tested.  If implemented, we would also need to add the `RUN_WRAPPER` to the bisect branch being developed.

**Comment by mikebentley15** [comment]
One challenge of this issue support is to call `MPI_Init()` and `MPI_Finalize()` within the FLiT framework before any state changes are made.  We would want to do this in a way to ensure that it still works if there is no support for MPI on the computer being used.  But if there is MPI, then we can safely call `MPI_Init()` and `MPI_Finalize()` as the first and last things respectively.

Perhaps it would  be a good idea to have a configuration option in the `flit-config.toml` called something like `enable_mpi`.  If this is used, we can have another internal run wrapper that will be automatically populated.  We would also need a configuration such as `mpi_procs`.  If those are set, then we know it is safe to assume we need to initialize and clean up the MPI environment.

**Comment by mikebentley15** [comment]
I have added the `RUN_WRAPPER`, but there was more necessary.  You enable the use of MPI in `flit-config.toml` with the `enable_mpi` variable in the `run` section.  I pull the compiler flags from `mpic++` automatically if the user has enabled MPI in this way.  As of now, it is working in my feature branch called `issue127-mpi-support`.

However, there is still a problem.  If you launch multiple copies of the same executable, then all of them will output the same information to the terminal and also output file(s).  This should be changed to only accept test results from rank 0 and only write out to files from rank 0.  This makes for a much larger change, but we can now use conditional compilation on the `__FLIT_USE_MPI__` macro variable.

As an aside, that variable should probably be renamed to `FLIT_USE_MPI`.

**Comment by mikebentley15** [comment]
For tests, we will need at least

- Compiling some flit tests with mpi enabled
- Running the tests with n = 1 and with n > 1
- Running tests with `--timing-loops -1` to make sure they only loop once
- Utilizing `RUN_WRAPPER`, perhaps the wrapper can simply be `echo` or something else simple

Documentation needs to address

- `RUN_WRAPPER` from within the `custom.mk` (although this is already documented in `custom.mk`)
- `enable_mpi` in `flit-config.toml`
- `mpirun_args` in `flit-config.toml`
- Test results are only captured from rank 0
- Auto timing loops will be replaced with looping once only, specify how many loops you want if one is not enough
- Only supports the `mpic++` and `mpirun` executables that are found on the `PATH`

**Comment by mikebentley15** [comment]
Actually about the timing loops, technically we should only have to disable that if we have more than one MPI process, not merely for compiling with MPI and initializing the environment.  So this restriction should be reduced.

**Comment by mikebentley15** [comment]
Automated tests have been delivered to the `issue127-mpi-support` branch to test the MPI support.  As a warning, it will fail if MPI is not on the system.  I should probably do something about that.  Perhaps skip the test and give a warning that the test has been skipped and why.

Much of the documentation has been updated on the `issue127-mpi-support` branch.  Almost done.

Finished at this point:
- Cuda support
- MPI support
- Writing test cases: this had some content but it is much more complete now
- FLiT configuration file: just added the MPI stuff

Still todo:
- Run wrapper and HPC support: Talk about RUN_WRAPPER and some examples of
  using it.  I will also want to perform some manual tests at least with
  Slurm to verify it works as expected.
- Compiling your tests: Talk about custom.mk and go through an example of
  getting it working on your code base (maybe Laghos and MFEM examples)

**Comment by mikebentley15** [comment]
The automated tests that have been implemented will fail to compile if the machine doing testing does not have MPI.  This is not good.  Instead, we need somehow to skip the MPI tests if there is no `mpirun` or `mpic++` in the path.  We will want to issue a warning that the MPI tests are skipped and why.

**Comment by mikebentley15** [comment]
The skipping of the automated tests has now been implemented.

There is one more thing to add to the documentation, and that is the dependencies if you choose to use MPI.  The dependencies are `mpirun` and `mpic++`.  If you have Ubuntu and want to use OpenMPI, then you would install them with

```bash
sudo apt install openmpi-bin libopenmpi-dev
```

**Comment by mikebentley15** [comment]
All documentation has been done along with automated tests.  It is now ready to be merged.

**Comment by mikebentley15** [comment]
Fixed by pull request #170
